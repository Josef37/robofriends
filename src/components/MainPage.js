import React from "react";
import Header from "../components/Header";
import CardList from "../components/CardList";
import SearchBox from "../components/SearchBox";
import Scroll from "../components/Scroll";
import ErrorBoundary from "../components/ErrorBoundary";
import "./MainPage.css";

class MainPage extends React.Component {
  componentDidMount() {
    this.props.onRequestRobots();
  }

  filterRobots = () => {
    const { robots, searchField } = this.props;
    return robots.filter(robot => {
      return searchField
        .toLowerCase()
        .split(" ")
        .every(subString => robot.name.toLowerCase().includes(subString));
    });
  };

  render() {
    const { isPending, onSearchChange } = this.props;

    return isPending ? (
      <h1>Loading</h1>
    ) : (
      <div className="tc">
        <Header />
        <SearchBox onSearchChange={onSearchChange} />
        <Scroll>
          <ErrorBoundary>
            <CardList robots={this.filterRobots()} />
          </ErrorBoundary>
        </Scroll>
      </div>
    );
  }
}

export default MainPage;
