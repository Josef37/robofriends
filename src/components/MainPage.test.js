import React from "react";
import { shallow } from "enzyme";
import MainPage from "./MainPage";

describe("MainPage smart component", () => {
  const mockInitialProps = {
    onRequestRobots: jest.fn(),
    robots: [],
    searchField: "",
    isPending: false
  };

  it("renders without crashing", () => {
    const wrapper = shallow(<MainPage {...mockInitialProps} />);
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("renders loading", () => {
    const mockProps = { ...mockInitialProps, isPending: true };
    const wrapper = shallow(<MainPage {...mockProps} />);
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("filters robots correctly", () => {
    const mockProps = {
      ...mockInitialProps,
      robots: [
        { id: 1, name: "Robot" },
        { id: 5, name: "not" }
      ],
      searchField: "bo"
    };
    const wrapper = shallow(<MainPage {...mockProps} />);
    expect(wrapper.instance().filterRobots()).toEqual([
      { id: 1, name: "Robot" }
    ]);
  });

  it("filters case insensitive", () => {
    const mockProps = {
      ...mockInitialProps,
      robots: [
        { id: 1, name: "Robot" },
        { id: 5, name: "Bot" }
      ],
      searchField: "boT"
    };
    const wrapper = shallow(<MainPage {...mockProps} />);
    expect(wrapper.instance().filterRobots()).toEqual(mockProps.robots);
  });

  it("filters every word", () => {
    const mockProps = {
      ...mockInitialProps,
      robots: [
        { id: 1, name: "Hello Bot" },
        { id: 5, name: "Hell Bot" }
      ],
      searchField: "hello ll ot"
    };
    const wrapper = shallow(<MainPage {...mockProps} />);
    expect(wrapper.instance().filterRobots()).toEqual([
      { id: 1, name: "Hello Bot" }
    ]);
  });
});
