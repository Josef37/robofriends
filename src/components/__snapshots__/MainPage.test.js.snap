// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage smart component renders loading 1`] = `
"<h1>
  Loading
</h1>"
`;

exports[`MainPage smart component renders without crashing 1`] = `
"<div className=\\"tc\\">
  <Header />
  <SearchBox onSearchChange={[undefined]} />
  <Scroll>
    <ErrorBoundary>
      <CardList robots={{...}} />
    </ErrorBoundary>
  </Scroll>
</div>"
`;
